sdv/single_table/base.py

```diff
@@ -42,3 +46,166 @@ def get_parameters(self):
+        keys.update(self._get_set_of_sequence_keys())
+        for key in sorted(keys):
+            if pd.isna(data[key]).any():
+                errors.append(f"Key column '{key}' contains missing values.")
+
+        return errors
+
+    @staticmethod
+    def _format_invalid_values_string(invalid_values):
+        invalid_values = sorted(invalid_values, key=lambda x: str(x))
+        if len(invalid_values) > 3:
+            return invalid_values[:3] + [f'+ {len(invalid_values) - 3} more']
+
+        return invalid_values
+
+    def _validate_key_values_are_unique(self, data):
+        errors = []
+        keys = self._get_primary_plus_alternate_keys()
+        for key in sorted(keys):
+            repeated_values = set(data[key][data[key].duplicated()])
+            if repeated_values:
+                repeated_values = self._format_invalid_values_string(repeated_values)
+                errors.append(f"Key column '{key}' contains repeating values: {repeated_values}")
+
+        return errors
+
+    def _validate_context_columns(self, data):
+        errors = []
+        context_column_names = self._data_processor._model_kwargs.get('context_columns')
+        if context_column_names:
+            sequence_key_names = sorted(self._get_set_of_sequence_keys())
+            for sequence_key, group in data.groupby(sequence_key_names):
+                if len(group.groupby(context_column_names)) != 1:
+                    invalid_context_cols = {}
+                    for context_column_name in context_column_names:
+                        values = set(group[context_column_name])
+                        if len(values) != 1:
+                            invalid_context_cols[context_column_name] = values
+
+                    errors.append(
+                        f'Context column(s) {invalid_context_cols} are changing '
+                        f'inside the sequence keys ({sequence_key_names}: {sequence_key}).'
+                    )
+
+        return errors
+
+    @staticmethod
+    def _is_numerical(value):
+        return pd.isna(value) | pd.api.types.is_float(value) | pd.api.types.is_integer(value)
+
+    @staticmethod
+    def _is_boolean(value):
+        return True if pd.isna(value) | (value is True) | (value is False) else False
+
+    def _validate_sdtype(self, sdtype, column, validation):
+        valid = column.apply(validation)
+        invalid_values = set(column[~valid])
+        if invalid_values:
+            invalid_values = self._format_invalid_values_string(invalid_values)
+            return [f"Invalid values found for {sdtype} column '{column.name}': {invalid_values}."]
+
+        return []
+
+    def _validate_column(self, column):
+        """Validate values of the column satisfy its sdtype properties."""
+        errors = []
+        sdtype = self.metadata._columns[column.name]['sdtype']
+
+        # boolean values must be True/False, None or missing values
+        # int/str are not allowed
```

Comment:
I think we can move this in the `docstring` or just remove it.

Reply:
If you don't have strong preferences, I'd prefer to leave it here. It's not that relevant to put in the docstrings, but I think it's useful for whoever is reading the code in the future (I, at least, was very confused what exactly each datatype means, and this comment avoids me having to check the utils file to figure it out).

Reply:
No strong preferences